### PR TITLE
initial haproxy-operator charm for review

### PR DESCRIPTION
**Pull request by dshcherb**
This charm provides means to deploy and operate haproxy.

An overview of how it can be used can be found here:
https://github.com/dshcherb/bundle-cockroachdb-ha#overview

### CODE REVIEW

**Comment by devec0** [comment]
Hey Dmitrii - whilst I'm not part of the group working on Operator, I figured some feedback might be useful as I have been co-maintaining a haproxy charm recently, and am very interested in Operator and am excited about these charms.

One thing about the promulgated haproxy charm is that it is tied solely to HTTP interfaces, so I really like that you've gone for a more generic TCP listener interface - this is a great move to improve the flexibility of the charm, as there are many use cases for haproxy beyond HTTP load balancing, so it's nice to support this. One thing that I have found, is that a lot of Juju ecosystem around load balancing is tied to HTTP, and specifically tied to the hacluster charm, which is a pattern I think it would be good to get away from. With a specific keepalived and haproxy charm that supports TCP interfaces for arbitrary services - we have a very flexible solution for load balancing, which is awesome.

In relation to this, I had a few points I think it would be good to consider (I am more than happy to help with these, if that would be helpful):

The TCP listener interface should be made more generic to include, rather than haproxy-specific strings for listener and backend options - general constructs which would be useful to any load balancer. For example, being able to set mode: http on the interface to enable HTTP-related checks and load balancing logic. Or being able to set 'http2' or 'timeout' options, which could be interpreted (or ignored!) by a specific load balancer charm based on whatever that charm has to do to configure the load balance to support those features I think is a much more flexible approach. The current approach ties the relation to haproxy, and requires the charm author relating to this haproxy charm to know haproxy quite well, in order to be able to add reverse-proxy support to their charm.

It would be good to be able to support different versions of haproxy, as the haproxy shipped with LTS Ubuntu does not always support modern features such as HTTP/2 - I know this charm is WiP though, so this is more a reminder that somewhere later it would be good to support this, but I think adding a generic interface to APT and Snap packaging generally for the Operator framework would be a good way to address this for many charms, rather than implementing logic for handling package installs within the charm itself.

It would also be great to consider how the charm might support TLS in the future. One of haproxy's features that is super useful is being able to TLS-terminate - so setting backend mode and port becomes even more important - because the mode of the backend becomes an important piece of data when deciding whether or not you can enable TLS-termination at the haproxy level - in order for that work you need to know if you can set HTTP mode on the listener/backends.

This is just some initial high-level stuff, but let me know if you want to discuss any further ideas about this charm.

Great work!

**Comment by dshcherb** [comment]
Hi James, thanks for the feedback!

> The TCP listener interface should be made more generic to include, rather than haproxy-specific strings for listener and backend options - general constructs which would be useful to any load balancer.

Yes, right now it is a bit haproxy-specific but probably it shouldn't be. In my view, there is always a trade-off between supporting the largest amount of features of a particular backend vs a generic approach.

I could come up with some abstractions for different kinds of load-balancers (say: haproxy, nginx, LVS) - they will have some common properties but some options presented by backend charms will be ignored depending on what is supported.

For example, the following information could be included for TCP load-balancing:

* backend server hostnames, ports and optional weights;
* algorithm (round robin, weighted round robin, least-connection, weighted least-connection, locality-based least-connection, locality-based least-connection with replication, destination-hashing, source-hashing etc.);
  * https://cbonte.github.io/haproxy-dconv/2.0/configuration.html#4.2-balance
  * https://docs.nginx.com/nginx/admin-guide/load-balancer/tcp-udp-load-balancer/#configuring-tcp-or-udp-load-balancing
  * http://manpages.ubuntu.com/manpages/trusty/man8/ipvsadm.8.html
* hash type (for calculating things like IP hashes for session stickiness);
* TCP health-check parameters (port, connection timeout, send_string, expect_string, send_binary, expect_binary)
  *  https://cbonte.github.io/haproxy-dconv/2.0/configuration.html#4.2-option%20tcp-check
* HTTP health-check parameters: port (might be different from the target TCP port), HTTP path (<hostname>:<port>/<path>), request type (e.g. GET), expected body, expected http status code, expected headers, interval (between retries), delay_before_retry, jitter, max_failures, passes (the number of required successes of a health-check to consider a backend healthy), fails (the number of required fails to consider a backend unhealthy - max number of retries), mandatory (do not allow connections to a backend until a first health-check completes), TLS CA to use for https connections;
* forwarding method (direct_routing, ipip, masquerading).

Given that options are lb-specific, the interface code can get quite heavy on mappings from abstract options to concrete ones.

> adding a generic interface to APT and Snap packaging generally for the Operator framework would be a good way to address this for many charms, rather than implementing logic for handling package installs within the charm itself.

This is also true - there is definitely a need for that. I think we need to do it sooner than later. In this case I am aiming at demonstrating the use of the framework in its current form as a primary goal.

> It would also be great to consider how the charm might support TLS in the future.

There are several aspects to that I think:

- TLS termination can be done on the haproxy frontend;
- each HTTP backend can have its own TLS termination and haproxy needs to have the relevant CA certificates and DNS servers configured to perform server certificate and hostname verification;
- SNI can be used at the haproxy side and there needs to be a way to determine hostname to backend mapping.

I am going to think of a way to address this as well in some way as TLS certificates may come from different sources for backend charms and the load-balancer charm itself (some people want auto-generated certs, some want the ones provided explicitly by them).

**Comment by devec0** [comment]
Definitely agree there is a balance to be struck. It is definitely not possible to cover all custom and implementation-specific configuration for all load balancers, however there are common themes to the configuration of reverse proxies which I feel can be represented, beyond port number. An example of this is the reverseproxy interface (reactive) - which I have been involved in developing. Beyond basic settings, it allows an "options" dictionary to be shared between related units, which allows for those implementation-specific options which may or not be supported by remote proxies or applications - this to me feels like a good trade-off between being implementation-specific and allowing flexibility, without getting lost in the weeds of various features supported by various proxies.

I think as a bare minimum schema, you could consider -
 * frontend port
 * backend port
 * backend_ip (this can be derived from peer information, so may not be needed as a relation field)
 * external_hostname (for SNI or vhost-based backend matching, can be None)
 * external_path (for path-based backend forwarding, i.e. a subdirectory for HTTP)
 * protocol (tcp, http, others?)
 * options -> a hash of extra options which may be implementation specific for a given proxy/LB

There are some more examples that could be considered in the reverseproxy interface, but I feel like the above is a good level to start with, without getting implementation specific.

Regarding the other points, I totally understand the intention of the charms, and appreciate the work that has gone into writing them! My comments around package and TLS handling are certainly things which can be handled down the track. My only desire in mentioning these things, is so that the potential impact on interface and charm design can be considered at this early stage of the charms, the bundle and the framework whilst decisions are still being made about them. Thanks again.

**Comment by dshcherb** [comment]
@devec0 a more load-balancer-independent (but TCP-specific) interface called `tcp-load-balancer` is now up for review here: https://github.com/canonical/operator/pull/216.

I updated this PR (https://github.com/canonical/haproxy-operator/pull/1/commits/b752fc8070b15f4fa554ca9028d24642437f073d) to support it and functionally tested it manually with the cockroachdb-operator charm (I left the change to cockroachdb-operator outside the [currently open PR](https://github.com/canonical/cockroachdb-operator/pull/1) in a branch https://github.com/dshcherb/cockroachdb-operator-1/tree/2020-04-10-tcp-load-balancer-spike so that the review can be focused on the base functionality).

There is still some work to do:

- adding HTTP health-check rendering;
- unit testing.

Please let me know what you think about what is there so far if you have time.

Thanks!

File: src/haproxy_instance_manager.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+            self.tcp_backend_manager.pools,
+            bind_addresses,
+        )
+
+        self._stored.set_default(is_started=False)
+        self.haproxy_conf_file = Path(f'/etc/haproxy/juju-{self.model.app.name}.cfg')
+
+    @property
+    def is_started(self):
+        return self._stored.is_started
```
Comment (relaxdiego) [inline src/haproxy_instance_manager.py:37]: Would there be a possibility that this stored state will go out of sync with the underlying reality? If so, would it be better to exclusively rely on `systemd` for the actual status? For example:

```python
try:
    subprocess.check_call(['systemctl', 'is-active', '--quiet', 'haproxy'])
    return True
except subprocess.CalledProcessError:
    return False
```

**Comment by chipaca** [comment]
can I be taken off the reviewers list please?

**Pull request closed without merging**